public/app/plugins/datasource/cloudwatch/datasource.ts

```diff
@@ -234,64 +234,77 @@ export class CloudWatchDatasource extends DataSourceApi<CloudWatchQuery, CloudWa
         statsQuery: (param.statsGroups?.length ?? 0) > 0 ?? false,
       };
     });
+
+    const dataFrames = increasingInterval({ startPeriod: 100, endPeriod: 1000, step: 300 }).pipe(
+      concatMap(_ => this.makeLogActionRequest('GetQueryResults', queryParams)),
+      repeat(),
+      share()
+    );
+
     let prevRecordsMatched: Record<string, number> = {};
+    const consecutiveFailedAttempts = dataFrames.pipe(
+      scan((acc, frames) => {
+        let newAcc = acc + 1;
+        for (const frame of frames) {
+          const recordsMatched = frame.meta?.custom?.['Statistics']['RecordsMatched'];
+          if (recordsMatched > (prevRecordsMatched[frame.refId!] ?? 0)) {
+            newAcc = 0;
+          }
+          prevRecordsMatched[frame.refId!] = recordsMatched;
+        }
 
-    return withTeardown(
-      this.makeLogActionRequest('GetQueryResults', queryParams).pipe(
-        expand((dataFrames, i) => {
-          const allFramesCompleted = dataFrames.every(
-            dataFrame => dataFrame.meta?.custom?.['Status'] === CloudWatchLogsQueryStatus.Complete
-          );
-          return allFramesCompleted
-            ? empty()
-            : this.makeLogActionRequest('GetQueryResults', queryParams).pipe(
-                map(frames => {
-                  let moreRecordsMatched = false;
-                  for (const frame of frames) {
-                    const recordsMatched = frame.meta?.custom?.['Statistics']['RecordsMatched'];
-                    if (recordsMatched > (prevRecordsMatched[frame.refId!] ?? 0)) {
-                      moreRecordsMatched = true;
-                    }
-                    prevRecordsMatched[frame.refId!] = recordsMatched;
-                  }
-                  const noProgressMade = i >= MAX_ATTEMPTS - 2 && !moreRecordsMatched;
-                  if (noProgressMade) {
-                    for (const frame of frames) {
-                      _.set(frame, 'meta.custom.Status', CloudWatchLogsQueryStatus.Complete);
-                    }
-                  }
+        return newAcc;
+      }, 0),
+      share()
+    );
```

Comment:
Outstanding work with this! It's been a while since I used `scan` but couldn't you pass a more complex object then `0` as the seed? That would eliminate the use of `let prevRecordsMatched: Record<string, number> = {};` I think

Reply:
```typescript
    const consecutiveFailedAttempts = dataFrames.pipe(
      scan(
        ({ failures, prevRecordsMatched }, frames) => {
          failures++;
          for (const frame of frames) {
            const recordsMatched = frame.meta?.custom?.['Statistics']['RecordsMatched'];
            if (recordsMatched > (prevRecordsMatched[frame.refId!] ?? 0)) {
              failures = 0;
            }
            prevRecordsMatched[frame.refId!] = recordsMatched;
          }

          return { failures, prevRecordsMatched };
        },
        { failures: 0, prevRecordsMatched: {} as Record<string, number> }
      ),
      map(({ failures }) => failures),
      share()
    );
```

Something like this?

Reply:
Yeah, that looks excellent ⭐